File: 程序员面试金典/1.2.判定是否互为字符重排.py
class Solution:
    def CheckPermutation(self, s1: str, s2: str) -> bool:
        cnt = [0] * 26

        if len(s1) != len(s2):
            return False

        for i in range(len(s1)):
            o1 = ord(s1[i]) - ord('a')
            o2 = ord(s2[i]) - ord('a')

            cnt[o1] += 1
            cnt[o2] -= 1

        for i in range(26):
            if cnt[i] != 0:
                return False

        return True

File: 程序员面试金典/1.2.判定是否互为字符重排.cpp
#include <algorithm>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    bool CheckPermutation(string s1, string s2)
    {
        unordered_map<char, int> mem1;
        unordered_map<char, int> mem2;

        for (char c : s1) {
            mem1[c] += 1;
        }

        for (char c : s2) {
            mem2[c] += 1;
            if (mem2[c] > mem1[c]) {
                return false;
            }
        }

        return mem1 == mem2;
    }
};